fix(kselect): kselect reskin [KHCP-8998] (#1832)
Kong/kongponents: docs/components/select.md

  <KSelect label="Label" :items="selectItems" />
</ClientOnly>

```html
<KSelect label="Label" :items="selectItems" />
```

### labelAttributes

Label attributes to be passed to underlying [KLabel component](/components/label#props).

<ClientOnly>
  <KSelect :label-attributes="{ info: 'I use the KLabel `info` prop' }" label="Label" :items="selectItems" />
</ClientOnly>

```html
<KSelect :label-attributes="{ info: 'I use the KLabel `info` prop' }" label="Label" :items="selectItems" />
```

### placeholder

Placeholder to be displayed when no item is selected.

<ClientOnly>
  <KSelect placeholder="Select service" :items="selectItemsUnselected" />
</ClientOnly>

```html
<KSelect placeholder="Select service" :items="selectItems" />
```

### clearable

When set to `true`, the clear icon will be displayed in front of selected item.

<ClientOnly>
  <KSelect clearable :items="selectItems" />
</ClientOnly>

```html
<KSelect clearable :items="selectItems" />
```

### help

Help text to be displayed below the select element.

<ClientOnly>
  <KSelect help="Helpful text." :items="selectItems" />
</ClientOnly>

```html
<KSelect help="Helpful text." :items="selectItems" />
```

### error

> When set to `true`, error styling will apply.

Reviewer: The input element has an `errorMessage` prop, but this component reuses the `help` prop? Seems confusing - do we have a standard?

Author: Added a quick and easy way to sanitize user input: https://github.com/Kong/kongponents/pull/1832/commits/42cd593c12afc33cf797146d3074c346daf35c59. Though I would argue that this would need to be a separate ticket.

<ClientOnly>
  <KSelect error help="You can use `help` prop for displaying error messages." :items="selectItems" />
</ClientOnly>

```html
<KSelect error help="You can use `help` prop for displaying error messages." :items="selectItems" />
```

### width

Use this prop to limit element's width.

<ClientOnly>
  <KSelect width="200" :items="selectItems" />
</ClientOnly>

```html
<KSelect width="200" :items="selectItems" />
```

### dropdownMaxHeight

Maximum height for dropdown container. Defaults to `300px`.

<ClientOnly>
  <KSelect dropdown-max-height="300" :items="selectItems" />
</ClientOnly>

```html
<KSelect dropdown-max-height="300" :items="selectItems" />
```

### dropdownFooterText

> Text to be displayed at the bottom of dropdown container. Can also be [slotted](#dropdown-footer-text).

Reviewer: ```suggestion
Text to be displayed at the bottom of the dropdown container. Can also be [slotted](#dropdown-footer-text).
```

<ClientOnly>
  <KSelect dropdown-footer-text="Helpful text in the dropdown." :items="selectItems" />
</ClientOnly>

```html
<KSelect dropdown-footer-text="Helpful text in the dropdown." :items="selectItems" />
```

### dropdownFooterTextPosition

> Defaults to `sticky`, but also accepts `static` value should you want text passed through `dropdownFooterText` prop to bi displayed at the bottom of dropdown container after all items.

Reviewer: ```suggestion
Defaults to `sticky`, but also accepts `static` value should you want text passed through `dropdownFooterText` prop to be displayed at the bottom of dropdown container after all items.
```

<ClientOnly>
  <KSelect dropdown-footer-text-position="static" dropdown-footer-text="Helpful text in the dropdown." :items="selectItems" />
</ClientOnly>

```html
<KSelect dropdown-footer-text-position="static" dropdown-footer-text="Helpful text in the dropdown." :items="selectItems" />
```

### enableFiltering

> By default, items passed to KSelect are not searchable. When `enableFiltering` prop is true, clicking on select element will turn it into input field, typing into which will filter matching items.

Reviewer: ```suggestion
By default, items passed to KSelect are not searchable. When `enableFiltering` prop is true, clicking on the select element will turn it into input field. You may filter items by typing in the input which will hide non-matching items.
```

<ClientOnly>
  <KSelect enable-filtering placeholder="Try searching for 'service a'" :items="selectItemsUnselected" />
</ClientOnly>

```html
<KSelect enable-filtering placeholder="Try searching for 'service a'" :items="selectItems" />
```

### filterFunction

> Custom function to filter items by.

Reviewer: ```suggestion
A custom function to perform item filtering.
```

<ClientOnly>
  <KSelect enable-filtering :filter-function="tagsFilter" placeholder="Try searching for 'dev' or 'prod'" :items="selectItemsUnselectedTagged" />
</ClientOnly>

```vue
<template>
  <KSelect enable-filtering :filter-function="tagsFilter" placeholder="Try searching for 'dev' or 'prod'" :items="selectItems" />
</template>

<script setup lang="ts">
import type { SelectItem, SelectFilterFunctionParams } from '@kong/kongponents'

const selectItems: SelectItem[] = [{
  label: 'Service A',
  value: 'a',
}, {
  label: 'Service B',
  value: 'b',
  tags: ['dev']
}, {
  label: 'Service F',
  value: 'f',
  disabled: true,
}, {
  label: 'Service A1',
  value: 'a1',
  group: 'Series 1',
}, {
  label: 'Service B1',
  value: 'b1',
  group: 'Series 1',
}, {
  label: 'Service A2',
  value: 'a2',
  group: 'Series 2',
  tags: ['prod']
}, {
  label: 'Service B2',
  value: 'b2',
  group: 'Series 2',
}]

const tagsFilter = (params: SelectFilterFunctionParams) => 
  params?.items?.filter((item: SelectItem) => 
  item.label?.toLowerCase().includes(params.query?.toLowerCase())
    || item.tags?.includes(params.query?.toLowerCase()))
</script>
```

:::tip TIP

> Should you need to disable the default filter function while still allowing filtering (for example when you want to handle filtering asynchronously by fetching from the API), you can pass a function that always returns boolean `true` to this prop, like so: 

Reviewer: ```suggestion
Should you need to disable the default filter function while still allowing filtering (for example when you want to handle filtering asynchronously by querying an API), you can provide a function that always returns boolean `true` to this prop, like so: 
```

```html
<KSelect
  enable-filtering
  :filter-function="() => true"
  :items="selectItems"
/>
```
:::

### reuseItemTemplate

> Whether content passed through [`item-template` slot](#itemtemplate) should be applied for selected item.

Reviewer: ```suggestion
Reuse the same display format provided via the [`item-template` slot](#itemtemplate) for the selected item.
```

<ClientOnly>
  <KSelect
    :items="selectItems"
    reuse-item-template
  >
    <template #item-template="{ item }">
      <KongIcon />
      {{ item?.label }}
    </template>
  </KSelect>
</ClientOnly>

```html
<KSelect
  :items="selectItems"
  reuse-item-template
>
  <template #item-template="{ item }">
    <KongIcon />
    {{ item?.label }}
  </template>
</KSelect>
```

### positionFixed

> Use fixed positioning of the popover to avoid content being clipped by parental boundaries. See [KPopover](/components/popover#positionfixed) docs for more information.

Reviewer: ```suggestion
Use fixed positioning for the select dropdown to avoid content being clipped by the boundaries of its parent container. See [KPopover](/components/popover#positionfixed) docs for more information.
```

### enableItemCreation

> When used in conjunction with `enableFiltering` set to `true`, KSelect will suggest adding a new value when filtering produces no results.

Reviewer: issue: we need to sanitize HTML and input to prevent things like this (same comment on the multiselect PR)
![image](https://github.com/Kong/kongponents/assets/2229946/2e285420-1929-4eb5-a712-5dda5aeb49b0)

Reviewer: If a user enters HTML or script tags when creating a new item, it's adding that HTML directly into the DOM (XSS).

You need to sanitize the user input to prevent this.

Author: Added a quick and easy way to sanitize user input: https://github.com/Kong/kongponents/pull/1832/commits/42cd593c12afc33cf797146d3074c346daf35c59. Though I would argue that this would need to be a separate ticket.

Reviewer: This looks good; you'll need this on KMultiselect too

<ClientOnly>
  <KSelect enable-item-creation enable-filtering placeholder="Try searching for 'service d'" :items="selectItemsUnselected" />
</ClientOnly>

```html
<KSelect enable-item-creation enable-filtering placeholder="Try searching for 'service d'" :items="selectItems" />
```

### loading

> Pass `true` to display loader instead of items in the dropdown. KSelect's `item` prop is reactive to changes by design, so that should you need to perform async item fetching/filtering you can execute that logic on host app's side and pass items back to KSelect. The example below utilizes the [`@query-change` event](#query-change) to simulate async items fetching behind the scenes.

Reviewer: ```suggestion
Pass `true` to display loader instead of items in the dropdown. KSelect's `item` prop is reactive to changes by design, so that should you need to perform async item fetching/filtering you can execute that logic within the host app and pass items back to KSelect. 

The example below utilizes the [`@query-change` event](#query-change) to simulate async item fetching behind the scenes.
```

const fetchAsyncItems = (): void => {
  asyncItemsLoading.value = true

  setTimeout(() => {
    // normally we would perform api fetching here
    // but for the sake of this example let's just filter items locally with some timeout
    const items = JSON.parse(JSON.stringify(asyncItemsInitial))
    asyncItems.value = items.filter((item: SelectItem) => item.label?.toLowerCase().includes(asyncItemsQuery.value?.toLowerCase()))

    asyncItemsLoading.value = false
  }, 2000)
}
</script>
```

### kpopAttributes

Attributes to be passed to underlying KPopover component. See [KPopover's props](/components/popover#props) for more info.

### HTML Attributes

#### required

KSelect will display a red dot next to label when `required` attribute is passed.

<ClientOnly>
  <KSelect
    label="Label"
    required
    :items="selectItems"
  />
</ClientOnly>

```html
<KSelect
  label="Label"
  required
  :items="selectItems"
/>
```

#### disabled

Use this attribute to disable interaction with the element.

<ClientOnly>
  <KSelect
    label="Disabled"
    disabled
    :items="selectItems"
  />
</ClientOnly>

```html
<KSelect
  label="Disabled"
  disabled
  :items="selectItems"
/>
```

## Slots

> ### label-tooltip

Reviewer: suggestion: Can we move `label-tooltip` below the other, more important slots?
